apk/apk.go

```diff
@@ -224,30 +239,72 @@ func createData(dataTgz io.Writer, info *nfpm.Info, sizep *int64) ([]byte, error
 	return dataDigest, nil
 }
 
-func createControl(controlTgz io.Writer, info *nfpm.Info, size int64) ([]byte, error) {
-	builderControl := createBuilderControl(info, size)
-	controlDigest, err := writeTgz(controlTgz, tarCut, builderControl, sha256.New())
+func createControl(controlTgz io.Writer, info *nfpm.Info, size int64, dataDigest []byte) ([]byte, error) {
+	builderControl := createBuilderControl(info, size, dataDigest)
+	controlDigest, err := writeTgz(controlTgz, tarCut, builderControl, sha1.New()) // nolint:gosec
 	if err != nil {
 		return nil, err
 	}
 	return controlDigest, nil
 }
 
-func combineToApk(target io.Writer, dataTgz, controlTgz io.Reader) error {
-	for _, tgz := range []io.Reader{controlTgz, dataTgz} {
+func createSignature(signatureTgz io.Writer, info *nfpm.Info, controlSHA1Digest []byte) error {
+	signatureBuilder := createSignatureBuilder(controlSHA1Digest, info)
+	// we don't actually need to produce a digest here, but writeTgz
+	// requires it so we just use SHA1 since it is already imported
+	_, err := writeTgz(signatureTgz, tarCut, signatureBuilder, sha1.New()) // nolint:gosec
+	return err
+}
+
+func createSignatureBuilder(digest []byte, info *nfpm.Info) func(*tar.Writer) error {
+	return func(tw *tar.Writer) error {
+		signature, err := sign.RSASignSHA1Digest(digest,
+			info.APK.SigningKeyFile, info.APK.SigningKeyPassphrase)
+		if err != nil {
+			return err
+		}
+
+		// needs to exist on the machine during installation: /etc/apk/keys/<keyname>.rsa.pub
+		keyname := info.APK.SigningKeyName
+		if keyname == "" {
+			addr, err := mail.ParseAddress(info.Maintainer)
+			if err != nil {
+				return errors.Wrap(err, "key name not send and unable to parse maintainer mail address")
```

Comment:
maybe just `unable to parse maintainer email address`?

Reply:
I wanted the user to know why I am parsing the maintainer email in the first place. I just spotted a type though (`send` instead of `set`)